minichain/mpt.py

```
@@ -0,0 +1,164 @@
+import json
+from typing import Optional, List
+
+def hash_data(data: bytes) -> bytes:
+    return hashlib.sha256(data).digest()
+
+def to_nibbles(key_hex: str) -> List[int]:
+    """Converts a hex string key into a list of integer nibbles (0-15)."""
+    try:
+        return [int(c, 16) for c in key_hex]
+    except ValueError:
+        raise ValueError(f"Invalid MPT key: '{key_hex}'. Keys must be valid hex strings.")
```

Ruff (0.15.14)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)